### ISSUE TITLE
Incorrect hydration 

### ISSUE THREAD

**Issue by ladal1** [mikro-orm/mikro-orm#5245]
### Describe the bug

Code leads to fake hydration of relation, meaning new loads from identity map return incorrect data (undefined instead of data).

We've seen this issue two times where identity map had incorrect data loaded (relation was not populated), this is the one I was able to replicate, I'm guessing it's an issue with query builder interaction with the identity map, but in the second case it might have been the OneToOne relation, but I wasn't able to completely diagnose that yet.

### Reproduction

https://gist.github.com/ladal1/9b27afcf195f10056c173a57569be3af

I've only quickly stitched it together so I might have once again made a mistake in it, but hopefully it's actually in there

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.1.0

### Node.js version

18.17.1

### Operating system

OS X 14.1.1

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Slack](https://join.slack.com/t/mikroorm/shared_invite/enQtNTM1ODYzMzM4MDk3LWM4ZDExMjU5ZDhmNjA2MmM3MWMwZmExNjhhNDdiYTMwNWM0MGY5ZTE3ZjkyZTMzOWExNDgyYmMzNDE1NDI5NjA).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
The QB usage here is not considered as partial loading apparently, that's the difference. If you would use `await orm.em.find(Test, {}, { fields: ['id'] })` instead it works correctly.

**Comment by ladal1** [comment]
Agreed, however we have much more complex query which can't really be efficiently replicated using the entity manager, and we have to either fork the em or do something else

**Comment by B4nan** [comment]
Oh wait a sec, that's an inverse side of 1:1 relation, if you want that to work with QB you need to enable `QueryFlag.AUTO_JOIN_ONE_TO_ONE_OWNER`, that's why it works with the EM and not with QB, as this behavior is opt-in with QB.

You can't populate the value because the FK is not present. Sounds like a wontfix to me, the population mechanism depends on this flag, but will take a closer look first.

**Comment by ladal1** [comment]
I expected it to not join on the query builder, the issue is that repeated find with explicit populate does not load it, which I feel like is actual issue

**Comment by B4nan** [comment]
It's really about this flag, if you disable it, you get the same problem with `em.find` and partial loading.

**Comment by B4nan** [comment]
I guess the problem is here, we handle the population conditionally based on that global option, but this needs to work based on the owning entity state (or better say the partial loading hint)

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/EntityLoader.ts#L325
